staging/src/k8s.io/apiextensions-apiserver/pkg/openapi/aggregator.go

```diff
@@ -0,0 +1,232 @@
+			"initial_static_spec": {
+				spec: staticSpec,
+			},
+		},
+		openAPIService:          openAPIService,
+		openAPIVersionedService: openAPIVersionedService,
+	}, nil
+}
+
+// openAPISpecInfo is used to store OpenAPI spec with its priority.
+// It can be used to sort specs with their priorities.
+type openAPISpecInfo struct {
+	// Name of a registered ApiService
+	name string
+
+	// Specification of this API Service. If null then the spec is not loaded yet.
+	spec *spec.Swagger
+	etag string
+}
+
+// buildOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
+func (s *specAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err error) {
+	specs := []openAPISpecInfo{}
+	for _, specInfo := range s.openAPISpecs {
+		if specInfo.spec == nil {
+			continue
+		}
+		specs = append(specs, *specInfo)
+	}
+	if len(specs) == 0 {
+		return &spec.Swagger{}, nil
+	}
+	for _, specInfo := range specs {
+		if specToReturn == nil {
+			specToReturn, err = aggregator.CloneSpec(specInfo.spec)
+			if err != nil {
+				return nil, err
+			}
+			continue
+		}
+		mergeSpecs(specToReturn, specInfo.spec)
+	}
+	// Add minimum required keys if missing, to properly serve the OpenAPI spec
+	// through apiextensions-apiserver HTTP handler. These keys will not be
+	// aggregated to top-level OpenAPI spec (only paths and definitions will).
+	// However these keys make the OpenAPI->proto serialization happy.
+	if specToReturn.Info == nil {
+		specToReturn.Info = &spec.Info{
+			InfoProps: spec.InfoProps{
+				Title: "Kubernetes",
+			},
+		}
+	}
+	if len(specToReturn.Swagger) == 0 {
+		specToReturn.Swagger = "2.0"
+	}
+	return specToReturn, nil
+}
+
+// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
+func (s *specAggregator) updateOpenAPISpec() error {
+	if s.openAPIService == nil || s.openAPIVersionedService == nil {
+		// openAPIVersionedService and deprecated openAPIService should be initialized together
+		if !(s.openAPIService == nil && s.openAPIVersionedService == nil) {
+			return fmt.Errorf("unexpected openapi service initialization error")
+		}
+		return nil
+	}
+	specToServe, err := s.buildOpenAPISpec()
+	if err != nil {
+		return err
+	}
+	// openAPIService.UpdateSpec and openAPIVersionedService.UpdateSpec read the same swagger spec
+	// serially and update their local caches separately. Both endpoints will have same spec in
+	// their caches if the caller is holding proper locks.
+	err = s.openAPIService.UpdateSpec(specToServe)
+	if err != nil {
+		return err
+	}
+	return s.openAPIVersionedService.UpdateSpec(specToServe)
+}
+
+// tryUpdatingServiceSpecs tries updating openAPISpecs map with specified specInfo, and keeps the map intact
+// if the update fails.
+func (s *specAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
+	orgSpecInfo, exists := s.openAPISpecs[specInfo.name]
```

Comment:
can we short-circuit in the following cases and avoid doing the expensive aggregation again?
* exists && orgSpecInfo == nil && specInfo == nil
* exists && orgSpecInfo != nil && specInfo != nil && orgSpecInfo.etag == specInfo.etag